WTG.Analyzers/Analyzers/PublicTupleType/PublicTupleTypeAnalyzer.cs

```diff
@@ -0,0 +1,303 @@
+
+						if (syntax != null && syntax.IsExternallyVisible() && !syntax.ImplementsAnInterface())
+						{
+							return true;
+						}
+
+						for (var i = 1; i < variables.Count; i++)
+						{
+							syntax = model.GetDeclaredSymbol(variables[0], cancellationToken);
+
+							if (syntax != null && !syntax.ImplementsAnInterface())
+							{
+								return true;
+							}
+						}
+					}
+
+					return false;
+				}
+			}
+
+			public override void VisitIndexerDeclaration(IndexerDeclarationSyntax node)
+			{
+				if (!IsOverride(node.Modifiers))
+				{
+					var syntax = model.GetDeclaredSymbol(node, cancellationToken);
+
+					if (syntax != null && syntax.IsExternallyVisible() && !syntax.ImplementsAnInterface())
+					{
+						VisitParameters(node.ParameterList.Parameters);
+						Visit(node.Type);
+					}
+				}
+			}
+
+			public override void VisitDelegateDeclaration(DelegateDeclarationSyntax node)
+			{
+				var syntax = model.GetDeclaredSymbol(node, cancellationToken);
+
+				if (syntax != null && syntax.IsExternallyVisible())
+				{
+					VisitParameters(node.ParameterList.Parameters);
+					Visit(node.ReturnType);
+				}
+			}
+
+			public override void VisitArrayType(ArrayTypeSyntax node)
+			{
+				Visit(node.ElementType);
+			}
+
+			public override void VisitPointerType(PointerTypeSyntax node)
+			{
+				Visit(node.ElementType);
+			}
+
+			public override void VisitGenericName(GenericNameSyntax node)
+			{
+				foreach (var arg in node.TypeArgumentList.Arguments)
+				{
+					Visit(arg);
+				}
+
+				Validate(node);
+			}
+
+			public override void VisitTupleType(TupleTypeSyntax node)
+			{
+				foreach (var element in node.Elements)
+				{
+					Visit(element.Type);
+				}
+
+				Validate(node);
+			}
+
+			public override void VisitIdentifierName(IdentifierNameSyntax node)
+			{
+				Validate(node);
+			}
+
+			public override void VisitQualifiedName(QualifiedNameSyntax node)
+			{
+				Validate(node);
+			}
+
+			static bool IsOverride(SyntaxTokenList modifiers)
+			{
+				foreach (var modifier in modifiers)
+				{
+					if (modifier.IsKind(SyntaxKind.OverrideKeyword))
+					{
+						return true;
+					}
+				}
+
+				return false;
+			}
+
+			void VisitParameters(SeparatedSyntaxList<ParameterSyntax> parameters)
+			{
+				foreach (var parameter in parameters)
+				{
+					Visit(parameter.Type);
+				}
+			}
+
+			void Validate(TypeSyntax node)
+			{
+				var symbol = model.GetSymbolInfo(node).Symbol;
+
+				if (symbol != null && symbol.Kind == SymbolKind.NamedType)
+				{
+					var type = (INamedTypeSymbol)symbol;
+
+					if (IsTupleLike(type))
+					{
+						report(Diagnostic.Create(
+							Rules.AvoidTupleTypesInPublicInterfacesRule,
+							node.GetLocation()));
+					}
+				}
+			}
+
+			static bool IsTupleLike(INamedTypeSymbol symbol)
+			{
+				if (symbol.IsTupleType)
+				{
+					return true;
+				}
+
+				switch (symbol.TypeKind)
+				{
+					case TypeKind.Class:
+						if (symbol.Name == "Tuple" &&
```

Comment:
Should we have `IsMatchAnyArity("System.Tuple")`?

Reply:
Possibly, but would it only do that for the inner most class if nested?

Reply:
We can build that bridge when we get to it.